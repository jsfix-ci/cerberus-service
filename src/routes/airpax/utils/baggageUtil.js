import * as pluralise from 'pluralise';
import { UNKNOWN_TEXT } from '../../../constants';

const getBaggageWeight = (baggage) => {
  if (!baggage?.weight) {
    return UNKNOWN_TEXT;
  }
  return baggage.weight;
};

const hasBaggage = (targetTask) => {
  return !!targetTask.movement.baggage;
};

const getBaggage = (targetTask) => {
  if (hasBaggage(targetTask)) {
    return targetTask.movement.baggage;
  }
  return null;
};

const getCheckedBags = (baggage, taskDetails = false) => {
  if ((!baggage || !baggage?.numberOfCheckedBags) && baggage?.numberOfCheckedBags !== 0) {
    return UNKNOWN_TEXT;
  }
  return pluralise.withCount(
    baggage.numberOfCheckedBags,
    '% checked bag',
    '% checked bags',
    (taskDetails ? 'None' : 'No checked bags'),
  );
};

const getNumberOfCheckedBags = (baggage) => {
  if ((!baggage || !baggage?.numberOfCheckedBags) && baggage?.numberOfCheckedBags !== 0) {
    return UNKNOWN_TEXT;
  }
  return baggage.numberOfCheckedBags;
};

const toFormattedCheckedBagsCount = (baggage) => {
  if ((!baggage || !baggage?.numberOfCheckedBags) && baggage?.numberOfCheckedBags !== 0) {
    return UNKNOWN_TEXT;
  }
  return pluralise.withCount(
    baggage.numberOfCheckedBags,
    '% bag',
    '% bags',
    'None',
  );
};

const BaggageUtil = {
  get: getBaggage,
  weight: getBaggageWeight,
  checked: getCheckedBags,
  checkedCount: getNumberOfCheckedBags,
  formatCheckedCount: toFormattedCheckedBagsCount,
};

export default BaggageUtil;

export { getBaggageWeight, getBaggage, getCheckedBags, getNumberOfCheckedBags };
