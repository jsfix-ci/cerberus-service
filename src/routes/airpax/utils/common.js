import lookup from 'country-code-lookup';
import { UNKNOWN_TEXT } from '../../../constants';

const convertToIso3Code = (iso2Code) => {
  if (!iso2Code) {
    return UNKNOWN_TEXT;
  }
  if (!lookup.byIso(iso2Code)) {
    return UNKNOWN_TEXT;
  }
  return lookup.byIso(iso2Code).iso3;
};

const Common = {
  iso3Code: convertToIso3Code,
};

export default Common;

export { convertToIso3Code };
