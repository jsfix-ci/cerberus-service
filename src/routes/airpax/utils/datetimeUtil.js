import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { UNKNOWN_TEXT } from '../../../constants';

dayjs.extend(utc);

const getDate = () => {
  return dayjs.utc().format();
};

const getFormattedDate = (date, dateFormat) => {
  if (!date) {
    return UNKNOWN_TEXT;
  }
  return dayjs.utc(date).format(dateFormat);
};

const toDateTimeList = (dateOne, dateTwo) => {
  return [dateOne, dateTwo];
};

const DateTimeUtil = {
  date: getDate,
  format: getFormattedDate,
  toList: toDateTimeList,
};

export default DateTimeUtil;

export { getDate, getFormattedDate, toDateTimeList };
