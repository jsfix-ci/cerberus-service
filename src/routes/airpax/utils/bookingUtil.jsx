import React from 'react';
import lookup from 'country-code-lookup';
import { getFormattedDate, toDateTimeList } from './datetimeUtil';
import calculateTimeDifference from '../../../utils/calculateDatetimeDifference';

import {
  UNKNOWN_TEXT,
  SHORT_DATE_FORMAT_ALT,
  STANDARD_HOUR_MINUTE_FORMAT,
  STANDARD_CARD_EXPIRY_FORMAT,
} from '../../../constants';

const getAgentLocation = (agent) => {
  if (!agent?.location) {
    return UNKNOWN_TEXT;
  }
  return agent.location;
};

const getAgentIata = (agent) => {
  if (!agent?.iata) {
    return UNKNOWN_TEXT;
  }
  return agent.iata;
};

const getAgentName = (agent) => {
  if (!agent?.name) {
    return UNKNOWN_TEXT;
  }
  return agent.name;
};

const hasAgent = (booking) => {
  return !!booking?.agent;
};

const getAgent = (booking) => {
  if (hasAgent(booking)) {
    return booking.agent;
  }
  return null;
};

const hasPaymentCard = (payment) => {
  return !!payment?.card;
};

const getPaymentCard = (payment) => {
  if (hasPaymentCard(payment)) {
    return payment.card;
  }
  return null;
};

const getCardExpiry = (payment) => {
  const paymentCard = getPaymentCard(payment);
  if (paymentCard?.expiry) {
    return getFormattedDate(paymentCard.expiry, STANDARD_CARD_EXPIRY_FORMAT);
  }
  return UNKNOWN_TEXT;
};

const getCardLastFourDigits = (payment) => {
  const paymentCard = getPaymentCard(payment);
  if (paymentCard?.number) {
    return paymentCard.number.substr(paymentCard.number.length - 4);
  }
  return UNKNOWN_TEXT;
};

const getPaymentAmount = (payment) => {
  if (!payment?.amount) {
    return UNKNOWN_TEXT;
  }
  return payment.amount;
};

const hasPayments = (booking) => {
  return !!booking?.payments?.length > 0;
};

const getPayments = (booking) => {
  if (hasPayments(booking)) {
    return booking.payments;
  }
  return null;
};

const getTicketType = (ticket) => {
  if (!ticket?.type) {
    return UNKNOWN_TEXT;
  }
  return ticket.type;
};

const getTicketNumber = (ticket) => {
  if (!ticket?.number) {
    return UNKNOWN_TEXT;
  }
  return ticket.number;
};

const hasTicket = (booking) => {
  return !!booking?.ticket;
};

const getTicket = (booking) => {
  if (hasTicket(booking)) {
    return booking.ticket;
  }
  return null;
};

const getBookingType = (booking) => {
  if (!booking?.type) {
    return UNKNOWN_TEXT;
  }
  return booking.type;
};

const getBookingCountryCode = (booking) => {
  if (!booking?.country) {
    return UNKNOWN_TEXT;
  }
  return booking.country;
};

const getBookingCountryName = (booking) => {
  const countryCode = getBookingCountryCode(booking);
  if (countryCode === UNKNOWN_TEXT) {
    return UNKNOWN_TEXT;
  }
  if (lookup.byIso(countryCode) === null) {
    return UNKNOWN_TEXT;
  }
  return lookup.byIso(countryCode).country;
};

const toBookingDateText = (booking) => {
  if (!booking?.bookedAt) {
    return UNKNOWN_TEXT;
  }
  return getFormattedDate(booking.bookedAt, SHORT_DATE_FORMAT_ALT);
};

const getBookingReference = (booking) => {
  if (!booking?.reference) {
    return UNKNOWN_TEXT;
  }
  return booking.reference;
};

const getCheckInAt = (booking) => {
  if (!booking?.checkInAt) {
    return UNKNOWN_TEXT;
  }
  return booking.checkInAt;
};

const toCheckInTimeText = (booking) => {
  const checkinPrefix = 'Check-in';
  if (!booking?.checkInAt) {
    return `${checkinPrefix} ${UNKNOWN_TEXT}`;
  }
  return `${checkinPrefix} ${getFormattedDate(booking.checkInAt, STANDARD_HOUR_MINUTE_FORMAT)}`;
};

const getBookedAt = (booking) => {
  return booking?.bookedAt;
};

const getBookedPriorToDeparture = (bookedAt, departureTime) => {
  const dateTimeList = toDateTimeList(bookedAt, departureTime);
  return calculateTimeDifference(dateTimeList);
};

const hasBooking = (targetTask) => {
  return !!targetTask?.movement?.booking;
};

const getBooking = (targetTask) => {
  if (hasBooking(targetTask)) {
    return targetTask.movement.booking;
  }
  return null;
};

const toPaymentsBlock = (booking) => {
  const payments = getPayments(booking);
  if (payments) {
    return payments.map((payment, index) => {
      return (
        <div key={index} className="bottom-border-thin govuk-!-margin-top-1 govuk-!-padding-bottom-1">
          <div className="font__bold">{getPaymentAmount(payment)}</div>
          <div className="font__light">
            Credit card ending {getCardLastFourDigits(payment)}, expiry {getCardExpiry(payment)}
          </div>
        </div>
      );
    });
  }
};

const BookingUtil = {
  get: getBooking,
  bookedAt: getBookedAt,
  checkInAt: getCheckInAt,
  bookedPrior: getBookedPriorToDeparture,
  toCheckInText: toCheckInTimeText,
  bookingRef: getBookingReference,
  toBookingText: toBookingDateText,
  countryCode: getBookingCountryCode,
  countryName: getBookingCountryName,
  bookingType: getBookingType,
  bookingTicket: getTicket,
  ticketNumber: getTicketNumber,
  ticketType: getTicketType,
  payments: getPayments,
  containsPayments: hasPayments,
  paymentAmount: getPaymentAmount,
  paymentCard: getPaymentCard,
  cardLastFourDigits: getCardLastFourDigits,
  cardExpiry: getCardExpiry,
  agent: getAgent,
  agentName: getAgentName,
  agentIata: getAgentIata,
  agentLocation: getAgentLocation,
  paymentsBlock: toPaymentsBlock,
};

export default BookingUtil;

export {
  toBookingDateText,
  getBookingReference,
  toCheckInTimeText,
  getBooking,
  getBookedAt,
  getCheckInAt,
  getBookedPriorToDeparture,
  getBookingCountryCode,
  getBookingCountryName,
  getBookingType,
  getTicket,
  getTicketNumber,
  getTicketType,
  getPayments,
  getPaymentAmount,
  getPaymentCard,
  getCardLastFourDigits,
  getCardExpiry,
  getAgent,
  getAgentName,
  getAgentIata,
  getAgentLocation,
  toPaymentsBlock,
};
