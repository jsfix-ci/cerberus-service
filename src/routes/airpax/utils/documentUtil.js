import moment from 'moment';
import lookup from 'country-code-lookup';
import {
  A_TITLE_CASE_TEXT,
  AN_TITLE_CASE_TEXT,
  AGO_TEXT,
  A_SMALL_TEXT,
  AN_SMALL_TEXT,
  UNKNOWN_TEXT,
  AFTER_TRAVEL_TEXT,
  BEFORE_TRAVEL_TEXT,
  SHORT_DATE_FORMAT_ALT,
} from '../../../constants';
import { formatField } from '../../../utils/formatField';
import { getFormattedDate } from './datetimeUtil';

const calculateExpiry = (passportExpiry, arrivalTime) => {
  const expiry = arrivalTime
    && passportExpiry !== 'Unknown'
    && `${arrivalTime},${moment(passportExpiry).format('YYYY-MM-DDTHH:mm:ss')}`;
  if (expiry) {
    return formatField('BOOKING_DATETIME', expiry)
      .split(', ')[1]
      .replace(BEFORE_TRAVEL_TEXT, AFTER_TRAVEL_TEXT)
      .replace(AGO_TEXT, BEFORE_TRAVEL_TEXT)
      .replace(A_SMALL_TEXT, A_TITLE_CASE_TEXT)
      .replace(AN_SMALL_TEXT, AN_TITLE_CASE_TEXT);
  }
  return UNKNOWN_TEXT;
};

const hasDocument = (person) => {
  return !!person?.document;
};

const getDocument = (person) => {
  if (hasDocument(person)) {
    return person.document;
  }
  return null;
};

const getDocumentExpiry = (document, taskDetails = false) => {
  const expiryPrefix = 'Expires';
  if (!document?.expiry) {
    return taskDetails ? UNKNOWN_TEXT : `${expiryPrefix} ${UNKNOWN_TEXT}`;
  }
  return taskDetails ? `${getFormattedDate(document?.expiry, SHORT_DATE_FORMAT_ALT)}` : `${expiryPrefix} ${getFormattedDate(document?.expiry, SHORT_DATE_FORMAT_ALT)}`;
};

const getDocumentValidity = (document, taskDetails = false) => {
  const validityPrefix = 'Valid from';
  if (!document?.validFrom) {
    return taskDetails ? UNKNOWN_TEXT : `${validityPrefix} ${UNKNOWN_TEXT}`;
  }
  return taskDetails ? `${getFormattedDate(document?.validFrom, SHORT_DATE_FORMAT_ALT)}` : `${validityPrefix} ${getFormattedDate(document?.validFrom, SHORT_DATE_FORMAT_ALT)}`;
};

const getDocumentType = (document) => {
  if (!document?.type) {
    return UNKNOWN_TEXT;
  }
  return document.type;
};

const getDocumentNumber = (document) => {
  if (!document?.number) {
    return UNKNOWN_TEXT;
  }
  return document.number;
};

const getDocumentName = (person) => {
  if (!person?.name) {
    return UNKNOWN_TEXT;
  }
  const firstNames = person?.name?.first || UNKNOWN_TEXT;
  const lastName = person?.name?.last || UNKNOWN_TEXT;
  return `${lastName.toUpperCase()}, ${firstNames}`;
};

// TODO finish implementation once data flows through
const getDocumentIdentification = (document) => {
  if (!document) {
    return UNKNOWN_TEXT;
  }
  return UNKNOWN_TEXT;
};

const getDocumentCountryOfIssue = (document, taskDetails = false) => {
  const countryOfIssuePrefix = 'Issued by';
  if (!document?.countryOfIssue) {
    return taskDetails ? UNKNOWN_TEXT : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
  }
  if (taskDetails) {
    return lookup.byIso(document.countryOfIssue) !== null
      ? `${lookup.byIso(document.countryOfIssue).country} (${document.countryOfIssue})`
      : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
  }

  return lookup.byIso(document.countryOfIssue) !== null
    ? `${countryOfIssuePrefix} ${document.countryOfIssue}`
    : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
};

const getDocumentCountryOfIssueCode = (document) => {
  if (!document?.countryOfIssue) {
    return UNKNOWN_TEXT;
  }
  return document.countryOfIssue;
};

const getDocumentNationality = (document, taskDetails = false) => {
  return getDocumentCountryOfIssue(document, taskDetails);
};

const getDocumentDOB = (document) => {
  if (!document?.dateOfBirth) {
    return UNKNOWN_TEXT;
  }
  return getFormattedDate(document?.dateOfBirth, SHORT_DATE_FORMAT_ALT);
};

const DocumentUtil = {
  get: getDocument,
  docExpiry: getDocumentExpiry,
  docValidity: getDocumentValidity,
  docType: getDocumentType,
  docNumber: getDocumentNumber,
  docName: getDocumentName,
  docIdentification: getDocumentIdentification,
  docCountry: getDocumentCountryOfIssue,
  docCountryCode: getDocumentCountryOfIssueCode,
  docNationality: getDocumentNationality,
  docDOB: getDocumentDOB,
  calculateExpiry,
};

export default DocumentUtil;

export {
  getDocument,
  getDocumentExpiry,
  getDocumentValidity,
  getDocumentType,
  getDocumentNumber,
  getDocumentName,
  getDocumentIdentification,
  getDocumentCountryOfIssue,
  getDocumentCountryOfIssueCode,
  getDocumentNationality,
  getDocumentDOB,
  calculateExpiry,
};
