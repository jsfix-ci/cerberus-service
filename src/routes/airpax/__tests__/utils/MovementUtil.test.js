import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { MovementUtil } from '../../utils';
import { UNKNOWN_TEXT, UNKNOWN_TIME_DATA } from '../../../../constants';

describe('MovementUtil', () => {
  let targetTaskMin;

  const intineries = [
    {
      'id': 'AC0850',
      'arrival': {
        'country': null,
        'location': 'YYZ',
        'time': '2018-10-03T13:05:00Z',
      },
      'departure': {
        'country': null,
        'location': 'CDG',
        'time': '2018-10-03T11:00:00Z',
      },
      'duration': 7500000,
    },
    {
      'id': 'BD0998',
      'arrival': {
        'country': null,
        'location': 'YYC',
        'time': '2018-10-03T18:16:00Z',
      },
      'departure': {
        'country': null,
        'location': 'YYZ',
        'time': '2018-10-03T16:05:00Z',
      },
      'duration': 7860000,
    },
    {
      'id': 'XZ0123',
      'arrival': {
        'country': null,
        'location': 'LHR',
        'time': '2018-10-03T21:19:20Z',
      },
      'departure': {
        'country': null,
        'location': 'YYC',
        'time': '2018-10-03T18:32:40Z',
      },
      'duration': 10000000,
    },
  ];

  beforeEach(() => {
    targetTaskMin = {
      movement: {
        id: 'AIRPAXTSV:CMID=9c19fe74233c057f25e5ad333672c3f9/2b4a6b5b08ea434880562d6836b1111',
        status: 'PRE_ARRIVAL',
        mode: 'AIR_PASSENGER',
        description: 'individual',
        journey: {
          id: 'BA103',
          arrival: {
            country: null,
            location: 'LHR',
            time: null,
          },
          departure: {
            country: null,
            location: 'FRA',
            time: '2020-08-07T17:15:00Z',
          },
          route: [
            'FRA',
            'LHR',
          ],
          itinerary: [
            {
              id: 'BA103',
              arrival: {
                country: null,
                location: 'LHR',
                time: null,
              },
              departure: {
                country: null,
                location: 'FRA',
                time: '2020-08-07T17:15:00Z',
              },
            },
          ],
          duration: 10000000,
        },
        flight: {
          departureStatus: 'DEPARTURE_CONFIRMED',
          number: 'BA103',
          operator: 'BA',
          seatNumber: null,
        },
      },
    };
  });

  const airlineCodesMin = [
    {
      'name': 'British Airways',
      'twolettercode': 'BA',
    },
    {
      'name': 'Air France',
      'twolettercode': 'AF',
    },
    {
      'name': 'Qantas',
      'twolettercode': 'QF',
    },
  ];

  it('should return the route if present', () => {
    const output = MovementUtil.movementRoute(targetTaskMin.movement.journey);
    expect(output).not.toBeUndefined();
    expect(output).not.toBeNull();
    expect(output).toEqual(targetTaskMin.movement.journey.route);
  });

  it('should return null if formatted seat number is not present', () => {
    const output = MovementUtil.seatNumber(targetTaskMin.movement.flight);
    expect(output).toEqual(`seat ${UNKNOWN_TEXT}`);
  });

  it('should return null if formatted seat number is present', () => {
    targetTaskMin.movement.flight.seatNumber = '15C';

    const output = MovementUtil.seatNumber(targetTaskMin.movement.flight);
    expect(output).toEqual('seat 15C');
  });

  it('should return a flight object if present', () => {
    const output = MovementUtil.movementFlight(targetTaskMin);
    expect(output).not.toBeUndefined();
    expect(output).not.toBeNull();
    expect(output).toEqual(targetTaskMin.movement.flight);
  });

  it('should return a journey object if present', () => {
    const output = MovementUtil.movementFlight(targetTaskMin);
    expect(output).not.toBeUndefined();
    expect(output).not.toBeNull();
    expect(output).toEqual(targetTaskMin.movement.flight);
  });

  it('should return a departure time if present', () => {
    const output = MovementUtil.departureTime(targetTaskMin.movement.journey);
    expect(output).toEqual(targetTaskMin.movement.journey.departure.time);
  });

  it('should return arrival time if present', () => {
    const output = MovementUtil.arrivalTime(targetTaskMin.movement.journey);
    expect(output).toEqual(targetTaskMin.movement.journey.arrival.time);
  });

  it('should return a formatted departure date and time if present', () => {
    const output = MovementUtil.formatDepartureTime(targetTaskMin.movement.journey);
    expect(output).toEqual('7 Aug 2020 at 17:15');
  });

  it('should return a formatted arrival date and time if present', () => {
    const output = MovementUtil.formatArrivalTime(targetTaskMin.movement.journey);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should return a departure location if present', () => {
    const output = MovementUtil.departureLoc(targetTaskMin.movement.journey);
    expect(output).toEqual(targetTaskMin.movement.journey.departure.location);
  });

  it('should return an arrival location if present', () => {
    const output = MovementUtil.arrivalLoc(targetTaskMin.movement.journey);
    expect(output).toEqual(targetTaskMin.movement.journey.arrival.location);
  });

  it('should return a flight number if present', () => {
    const output = MovementUtil.flightNumber(targetTaskMin.movement.flight);
    expect(output).toEqual(targetTaskMin.movement.flight.number);
  });

  it('should return the airline operator if present', () => {
    const output = MovementUtil.airlineOperator(targetTaskMin.movement.flight);
    expect(output).toEqual(targetTaskMin.movement.flight.operator);
  });

  it('should render the departure status if present', () => {
    const tree = renderer.create(MovementUtil.status(targetTaskMin)).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the departure status and details if present', () => {
    const tree = renderer.create(MovementUtil.status(targetTaskMin, true)).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should return the movement type if present', () => {
    const output = MovementUtil.movementType(targetTaskMin);
    expect(output).toEqual('Passenger');
  });

  it('should return the movement description text', () => {
    const output = MovementUtil.description(targetTaskMin);
    expect(output).toEqual('Single passenger');
  });

  it('should return the movement description text for multiple persons', () => {
    targetTaskMin.movement.description = 'group';

    targetTaskMin.movement.person = {
      name: {
        first: 'Isaiah',
        last: 'Ford',
        full: 'Isaiah Ford',
      },
      role: 'PASSENGER',
      dateOfBirth: '1966-05-13T00:00:00Z',
      gender: 'M',
      nationality: 'GBR',
      document: null,
    };

    targetTaskMin.movement.otherPersons = [
      {
        name: {
          first: 'Isaiah',
          last: 'Ford',
          full: 'Isaiah Ford',
        },
        role: 'PASSENGER',
        dateOfBirth: '1966-05-13T00:00:00Z',
        gender: 'M',
        nationality: 'GBR',
        document: null,
      },
    ];

    const output = MovementUtil.description(targetTaskMin);
    expect(output).toEqual('In group of 2');
  });

  it('should convert airline code to airline name', () => {
    const GIVEN_BA = MovementUtil.airlineName('BA', airlineCodesMin);
    const GIVEN_AF = MovementUtil.airlineName('AF', airlineCodesMin);
    const GIVEN_QF = MovementUtil.airlineName('QF', airlineCodesMin);

    expect(GIVEN_BA).toEqual('British Airways');
    expect(GIVEN_AF).toEqual('Air France');
    expect(GIVEN_QF).toEqual('Qantas');
  });

  it('should return a formatted string from the departure location code when location is present', () => {
    const outcome = MovementUtil.formatLoc(targetTaskMin.movement.journey.departure.location);
    expect(outcome).toEqual('Frankfurt, Germany');
  });

  it('should return unknown when departure location code is null', () => {
    targetTaskMin.movement.journey.departure.location = null;
    const outcome = MovementUtil.formatLoc(targetTaskMin.movement.journey.departure.location);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when departure location code is undefined', () => {
    targetTaskMin.movement.journey.departure.location = undefined;
    const outcome = MovementUtil.formatLoc(targetTaskMin.movement.journey.departure.location);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when departure location code is empty', () => {
    targetTaskMin.movement.journey.departure.location = '';
    const outcome = MovementUtil.formatLoc(targetTaskMin.movement.journey.departure.location);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when departure location code is invalid string', () => {
    targetTaskMin.movement.journey.departure.location = 'unknown';
    const outcome = MovementUtil.formatLoc(targetTaskMin.movement.journey.departure.location);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return a formatted string from the arrival location code when location is present', () => {
    const outcome = MovementUtil.formatLoc(targetTaskMin.movement.journey.arrival.location);
    expect(outcome).toEqual('London, United Kingdom');
  });

  it('should return unknown when arrival location code is null', () => {
    targetTaskMin.movement.journey.arrival.location = null;
    const outcome = MovementUtil.formatLoc(targetTaskMin.movement.journey.arrival.location);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when arrival location code is undefined', () => {
    targetTaskMin.movement.journey.arrival.location = undefined;
    const outcome = MovementUtil.formatLoc(targetTaskMin.movement.journey.arrival.location);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when arrival location code is empty', () => {
    targetTaskMin.movement.journey.arrival.location = '';
    const outcome = MovementUtil.formatLoc(targetTaskMin.movement.journey.arrival.location);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when arrival location code is invalid string', () => {
    targetTaskMin.movement.journey.arrival.location = 'unknown';
    const outcome = MovementUtil.formatLoc(targetTaskMin.movement.journey.arrival.location);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return the flight duration when present', () => {
    const outcome = MovementUtil.flightDuration(targetTaskMin.movement.journey);
    expect(outcome).toEqual(targetTaskMin.movement.journey.duration);
  });

  it('should return unknown when flight duration when duration is null', () => {
    targetTaskMin.movement.journey.duration = null;
    const outcome = MovementUtil.flightDuration(targetTaskMin.movement.journey);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when flight duration when duration is undefined', () => {
    targetTaskMin.movement.journey.duration = undefined;
    const outcome = MovementUtil.flightDuration(targetTaskMin.movement.journey);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when flight duration when duration is empty', () => {
    targetTaskMin.movement.journey.duration = '';
    const outcome = MovementUtil.flightDuration(targetTaskMin.movement.journey);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return the flight time flight duration is present', () => {
    const outcome = MovementUtil.formatFlightTime(targetTaskMin.movement.journey);
    expect(outcome).toEqual('2h 46m');
  });

  it('should return unknown when flight duration is null', () => {
    targetTaskMin.movement.journey.duration = null;
    const outcome = MovementUtil.formatFlightTime(targetTaskMin.movement.journey);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when flight duration is undefined', () => {
    targetTaskMin.movement.journey.duration = undefined;
    const outcome = MovementUtil.formatFlightTime(targetTaskMin.movement.journey);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when flight duration is empty', () => {
    targetTaskMin.movement.journey.duration = '';
    const outcome = MovementUtil.formatFlightTime(targetTaskMin.movement.journey);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when flight duration is invalid', () => {
    targetTaskMin.movement.journey.duration = 'A';
    const outcome = MovementUtil.formatFlightTime(targetTaskMin.movement.journey);
    expect(outcome).toEqual(UNKNOWN_TEXT);
  });

  it('should convert the given time in milliseconds to a time object', () => {
    const given = '10000000';

    const expected = { h: 2, m: 46, s: 40 };

    const output = MovementUtil.flightTimeObject(given);
    expect(output).toEqual(expected);
  });

  it('should return a time object when given is 0', () => {
    const expected = { h: 0, m: 0, s: 0 };
    const output = MovementUtil.flightTimeObject(0);
    expect(output).toEqual(expected);
  });

  it('should return a time object when given is string representation of 0', () => {
    const expected = { h: 0, m: 0, s: 0 };
    const output = MovementUtil.flightTimeObject('0');
    expect(output).toEqual(expected);
  });

  it('should return a time object with unknown h. m & s when given is null', () => {
    const given = null;
    const output = MovementUtil.flightTimeObject(given);
    expect(output).toEqual(UNKNOWN_TIME_DATA);
  });

  it('should return a time object with unknown h. m & s when given is undefined', () => {
    const given = undefined;
    const output = MovementUtil.flightTimeObject(given);
    expect(output).toEqual(UNKNOWN_TIME_DATA);
  });

  it('should return a time object with unknown h. m & s when given is empty', () => {
    const output = MovementUtil.flightTimeObject('');
    expect(output).toEqual(UNKNOWN_TIME_DATA);
  });

  it('should return a time object with unknown h. m & s when given is a mixture of numbers & characters', () => {
    const given = '3600-00000A';
    const output = MovementUtil.flightTimeObject(given);
    expect(output).toEqual(UNKNOWN_TIME_DATA);
  });

  it('should return the itenerary flight number', () => {
    const expected = 'BA103';
    const output = MovementUtil.itinFlightNumber(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual(expected);
  });

  it('should return unknown when the itenerary flight number is null', () => {
    targetTaskMin.movement.journey.itinerary[0].id = null;
    const output = MovementUtil.itinFlightNumber(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when the itenerary flight number is undefined', () => {
    targetTaskMin.movement.journey.itinerary[0].id = undefined;
    const output = MovementUtil.itinFlightNumber(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should return unknown when the itenerary flight number is an empty string', () => {
    targetTaskMin.movement.journey.itinerary[0].id = '';
    const output = MovementUtil.itinFlightNumber(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should return the itinerary departure country code', () => {
    targetTaskMin.movement.journey.itinerary[0].departure.country = 'GB';
    const expected = 'GB';
    const output = MovementUtil.itinDepartureCountryCode(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual(expected);
  });

  it('should return the country code using departure location when the departure country code is null', () => {
    const output = MovementUtil.itinDepartureCountryCode(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual('DE');
  });

  it('should return the country code using departure location when the departure country code is undefined', () => {
    targetTaskMin.movement.journey.itinerary[0].departure.country = undefined;
    const output = MovementUtil.itinDepartureCountryCode(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual('DE');
  });

  it('should return the country code using departure location when the departure country code is an empty string', () => {
    targetTaskMin.movement.journey.itinerary[0].departure.country = '';
    const output = MovementUtil.itinDepartureCountryCode(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual('DE');
  });

  it('should return unknown when the country code can not be determined using the departure location', () => {
    targetTaskMin.movement.journey.itinerary[0].departure.location = null;
    targetTaskMin.movement.journey.itinerary[0].departure.country = '';
    const output = MovementUtil.itinDepartureCountryCode(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should return the itinerary arrival country code', () => {
    targetTaskMin.movement.journey.itinerary[0].arrival.country = 'FR';
    const expected = 'FR';
    const output = MovementUtil.itinArrivalCountryCode(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual(expected);
  });

  it('should return the country code using the arrival location when arrival country code is null', () => {
    const output = MovementUtil.itinArrivalCountryCode(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual('GB');
  });

  it('should return the country code using the arrival location when arrival country code is undefined', () => {
    targetTaskMin.movement.journey.itinerary[0].arrival.country = undefined;
    const output = MovementUtil.itinArrivalCountryCode(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual('GB');
  });

  it('should return the country code using the arrival location when arrival country code is an empty string', () => {
    targetTaskMin.movement.journey.itinerary[0].arrival.country = '';
    const output = MovementUtil.itinArrivalCountryCode(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual('GB');
  });

  it('should return unknown when the country code can not be determined using the arrival location', () => {
    targetTaskMin.movement.journey.itinerary[0].arrival.location = null;
    targetTaskMin.movement.journey.itinerary[0].arrival.country = '';
    const output = MovementUtil.itinArrivalCountryCode(targetTaskMin.movement.journey.itinerary[0]);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  describe('toItineraryBlock', () => {
    const expected = ['Unknown', '3 hours later', '17 minutes later'];
    for (let i = 0; i < intineries.length; i += 1) {
      it('should calculate time difference between flight legs', () => {
        // As the first leg has nothing prior to it, unknown will be returned however not rendered.
        render(MovementUtil.itinRelativeTime(i, intineries[i], intineries));
        expect(screen.getByText(expected[i])).toBeInTheDocument();
      });
    }
  });
});
