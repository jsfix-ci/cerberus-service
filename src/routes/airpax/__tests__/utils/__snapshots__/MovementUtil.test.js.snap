// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovementUtil should render the departure status and details if present 1`] = `
Array [
  <span>
    Departure confirmed
     
  </span>,
  <strong
    className="hods-tag  airpax-status airpax-status__green"
  >
    DC
  </strong>,
]
`;

exports[`MovementUtil should render the departure status if present 1`] = `
<strong
  className="hods-tag  airpax-status airpax-status__green"
>
  DC
</strong>
`;
